"""
QQBotå¾®æœåŠ¡æµ‹è¯•å®¢æˆ·ç«¯
"""
import asyncio
import json
import sys
from pathlib import Path
from typing import Optional, Tuple
import consul as consul_lib
import grpc
from loguru import logger

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ° Python è·¯å¾„
sys.path.insert(0, str(Path(__file__).parent.parent))

from src.qqbot import qqbot_pb2, qqbot_pb2_grpc


class ConsulServiceDiscovery:
    """ConsulæœåŠ¡å‘ç°"""

    def __init__(self, consul_host: str = "localhost:8500"):
        # è§£æä¸»æœºå’Œç«¯å£
        if ":" in consul_host:
            host_str, port_str = consul_host.split(":", 1)
            port = int(port_str)
        else:
            host_str = consul_host
            port = 8500

        self.client = consul_lib.Consul(
            host=host_str,
            port=port,
            scheme="http",
            verify=False
        )
        self.consul_host = consul_host

    async def discover_service(self, service_name: str) -> Optional[Tuple[str, int]]:
        """
        ä»Consulå‘ç°æœåŠ¡

        Args:
            service_name: æœåŠ¡åç§°

        Returns:
            (host, port) å…ƒç»„ï¼Œå¦‚æœæœªæ‰¾åˆ°åˆ™è¿”å› None
        """
        try:
            logger.info(f"ğŸ” åœ¨ Consul {self.consul_host} ä¸­æŸ¥æ‰¾æœåŠ¡: {service_name}")

            # è·å–å¥åº·çš„æœåŠ¡å®ä¾‹
            index, nodes = self.client.health.service(
                service=service_name,
                passing=True
            )

            if not nodes:
                logger.warning(f"âš ï¸  æœªæ‰¾åˆ°å¥åº·çš„æœåŠ¡å®ä¾‹: {service_name}")
                return None

            # é€‰æ‹©ç¬¬ä¸€ä¸ªå¥åº·çš„å®ä¾‹
            node = nodes[0]
            service_info = node.get('Service', {})

            address = service_info.get('Address', '')
            port = service_info.get('Port', 0)

            # å¦‚æœæœåŠ¡åœ°å€æ˜¯ç©ºå­—ç¬¦ä¸²ï¼Œä½¿ç”¨èŠ‚ç‚¹çš„åœ°å€
            if not address:
                address = node.get('Node', {}).get('Address', '')

            logger.info(f"âœ… å‘ç°æœåŠ¡: {service_name} -> {address}:{port}")

            return address, port

        except Exception as e:
            logger.error(f"âŒ æœåŠ¡å‘ç°å¤±è´¥: {e}")
            return None


class QQBotClient:
    """åŸºäºConsulå‘ç°çš„QQBotå®¢æˆ·ç«¯"""

    def __init__(self, consul_host: str = "localhost:8500", service_name: str = "ew.qbot"):
        self.consul_host = consul_host
        self.service_name = service_name
        self.service_discovery = ConsulServiceDiscovery(consul_host)
        self.channel = None
        self.stub = None

    async def connect_via_consul(self) -> bool:
        """
        é€šè¿‡Consulå‘ç°å¹¶è¿æ¥åˆ°æœåŠ¡

        Returns:
            æ˜¯å¦è¿æ¥æˆåŠŸ
        """
        try:
            # 1. ä»Consulå‘ç°æœåŠ¡
            service_info = await self.service_discovery.discover_service(self.service_name)

            if not service_info:
                logger.error(f"âŒ æ— æ³•åœ¨Consulä¸­æ‰¾åˆ°æœåŠ¡: {self.service_name}")
                return False

            address, port = service_info
            target = f"{address}:{port}"

            # 2. è¿æ¥åˆ°gRPCæœåŠ¡
            logger.info(f"ğŸ”— è¿æ¥åˆ° gRPC æœåŠ¡: {target}")

            self.channel = grpc.aio.insecure_channel(target)
            self.stub = qqbot_pb2_grpc.QQBotServiceStub(self.channel)

            # 3. æµ‹è¯•è¿æ¥
            try:
                response = await asyncio.wait_for(
                    self.stub.HealthCheck(qqbot_pb2.HealthCheckRequest()),
                    timeout=5.0
                )

                if response.status == 200:
                    logger.info(f"âœ… è¿æ¥æˆåŠŸ: {response.message}")
                    return True
                else:
                    logger.error(f"âŒ æœåŠ¡ä¸å¥åº·: {response.message}")
                    return False

            except asyncio.TimeoutError:
                logger.error(f"â° è¿æ¥è¶…æ—¶: {target}")
                return False

        except Exception as e:
            logger.error(f"ğŸ’¥ é€šè¿‡Consulè¿æ¥å¤±è´¥: {e}")
            return False

    async def send_message(self, target_id: str, target_type: str, content_data: dict,
                          sender_id: str = "test_client") -> Optional[dict]:
        """
        å‘é€QQæ¶ˆæ¯

        Args:
            target_id: ç›®æ ‡IDï¼ˆQQå·æˆ–ç¾¤å·ï¼‰
            target_type: ç›®æ ‡ç±»å‹: "user" æˆ– "group"
            content_data: æ¶ˆæ¯å†…å®¹æ•°æ®
            sender_id: å‘é€è€…ID

        Returns:
            å‘é€ç»“æœå­—å…¸
        """
        if not self.stub:
            logger.error("âŒ æœªè¿æ¥åˆ°æœåŠ¡")
            return None

        try:
            # æ„å»ºæ¶ˆæ¯å†…å®¹
            content = qqbot_pb2.MessageContent()

            # æ ¹æ®å†…å®¹ç±»å‹è®¾ç½®å­—æ®µ
            if isinstance(content_data, str):
                content.text = content_data
            elif isinstance(content_data, dict):
                if content_data.get("type") == "MessageArray":
                    # æ„å»ºæ¶ˆæ¯æ®µåˆ—è¡¨
                    segments = []
                    for segment_data in content_data.get("data", []):
                        segment = qqbot_pb2.MessageSegment()
                        if isinstance(segment_data, dict):
                            segment.type = segment_data.get("type", "")
                            if isinstance(segment_data.get("data"), dict):
                                segment.data = json.dumps(segment_data["data"], ensure_ascii=False)
                            elif segment_data.get("data") is not None:
                                segment.data = json.dumps({"value": segment_data["data"]}, ensure_ascii=False)
                        elif isinstance(segment_data, str):
                            segment.type = "Text"
                            segment.data = json.dumps({"text": segment_data}, ensure_ascii=False)
                        segments.append(segment)

                    content.segments.extend(segments)
                else:
                    # å•ä¸ªæ¶ˆæ¯æ®µ
                    content.segment.type = content_data.get("type", "")
                    data = content_data.get("data", {})
                    if isinstance(data, dict):
                        content.segment.data = json.dumps(data, ensure_ascii=False)
                    elif data is not None:
                        content.segment.data = json.dumps({"value": data}, ensure_ascii=False)

            # æ„å»ºè¯·æ±‚
            request = qqbot_pb2.SendMessageRequest(
                target_id=str(target_id),
                target_type=target_type,
                content=content,
                sender_id=sender_id,
                metadata={
                    "client": "test_client",
                    "consul_host": self.consul_host
                }
            )

            logger.info(f"ğŸ“¨ å‘é€æ¶ˆæ¯åˆ°: {target_type} {target_id}")

            # å‘é€è¯·æ±‚
            response = await self.stub.SendMessage(request)

            result = {
                "success": response.status == 200,
                "status_code": response.status,
                "message": response.message
            }

            if response.data:
                try:
                    data = json.loads(response.data)
                    result.update(data)
                except:
                    result['raw_data'] = response.data

            if result['success']:
                logger.info(f"âœ… æ¶ˆæ¯å‘é€æˆåŠŸ!")
            else:
                logger.error(f"âŒ æ¶ˆæ¯å‘é€å¤±è´¥: {result['message']}")

            return result

        except Exception as e:
            logger.error(f"ğŸ’¥ å‘é€æ¶ˆæ¯å¤±è´¥: {e}")
            return None

    async def health_check(self) -> Optional[dict]:
        """å¥åº·æ£€æŸ¥"""
        if not self.stub:
            logger.error("âŒ æœªè¿æ¥åˆ°æœåŠ¡")
            return None

        try:
            response = await self.stub.HealthCheck(qqbot_pb2.HealthCheckRequest())

            result = {
                "status_code": response.status,
                "message": response.message
            }

            if response.data:
                try:
                    data = json.loads(response.data)
                    result.update(data)
                    logger.info(f"ğŸ“Š å¥åº·çŠ¶æ€: {response.message}")
                except:
                    logger.info(f"ğŸ“‹ åŸå§‹æ•°æ®: {response.data[:200]}...")

            return result

        except Exception as e:
            logger.error(f"âŒ å¥åº·æ£€æŸ¥å¤±è´¥: {e}")
            return None

    async def close(self):
        """å…³é—­è¿æ¥"""
        if self.channel:
            await self.channel.close()
            logger.info("ğŸ”Œ è¿æ¥å·²å…³é—­")


async def main():
    """ä¸»æµ‹è¯•å‡½æ•°"""
    import argparse

    parser = argparse.ArgumentParser(description="QQBotå¾®æœåŠ¡æµ‹è¯•å®¢æˆ·ç«¯")
    parser.add_argument("--consul", default="dc-a588.local:8500", help="ConsulæœåŠ¡å™¨åœ°å€")
    parser.add_argument("--service", default="ew.qbot", help="æœåŠ¡åç§°")
    parser.add_argument("--target-id", default=3222087513, help="ç›®æ ‡IDï¼ˆQQå·æˆ–ç¾¤å·ï¼‰")
    parser.add_argument("--target-type", choices=["user", "group"], default="user", help="ç›®æ ‡ç±»å‹")
    parser.add_argument("--content-type", choices=["text", "face", "image", "complex"], default="text",
                       help="æ¶ˆæ¯ç±»å‹")

    args = parser.parse_args()

    # é…ç½®æ—¥å¿—
    logger.remove()
    logger.add(
        sys.stdout,
        format="<green>{time:HH:mm:ss}</green> | <level>{level: <8}</level> | <level>{message}</level>",
        level="INFO",
        colorize=True
    )

    client = QQBotClient(
        consul_host=args.consul,
        service_name=args.service
    )

    try:
        # è¿æ¥æœåŠ¡
        logger.info(f"ğŸš€ é€šè¿‡ Consul å‘ç°æœåŠ¡: {args.service}")
        connected = await client.connect_via_consul()

        if not connected:
            logger.error("âŒ æ— æ³•è¿æ¥åˆ°æœåŠ¡ï¼Œæµ‹è¯•ç»ˆæ­¢")
            return

        print("\n" + "="*60)
        print("QQBot å¾®æœåŠ¡æµ‹è¯•")
        print("="*60)

        # å¥åº·æ£€æŸ¥
        logger.info("\n1. ğŸ©º å¥åº·æ£€æŸ¥...")
        health_result = await client.health_check()

        if health_result and health_result.get('status_code') == 200:
            logger.info("âœ… æœåŠ¡å¥åº·")
        else:
            logger.warning("âš ï¸  æœåŠ¡å¯èƒ½ä¸å¥åº·")

        # æ„å»ºæ¶ˆæ¯å†…å®¹
        content_data = None

        if args.content_type == "text":
            content_data = f"ã€QQBotæµ‹è¯•æ¶ˆæ¯ã€‘\næ—¶é—´: {asyncio.get_event_loop().time():.2f}\nè¿™æ˜¯ä¸€æ¡æµ‹è¯•æ¶ˆæ¯ã€‚\næœåŠ¡å‘ç°: {args.service} via {args.consul}"

        elif args.content_type == "face":
            content_data = {
                "type": "Face",
                "data": 14  # è¡¨æƒ…ID
            }

        elif args.content_type == "image":
            content_data = {
                "type": "Image",
                "data": {
                    "url": "https://uploadstatic.mihoyo.com/contentweb/20210804/2021080419123130780.png",
                    "summary": "æµ‹è¯•å›¾ç‰‡"
                }
            }

        elif args.content_type == "complex":
            content_data = {
                "type": "MessageArray",
                "data": [
                    "å–µå–µå–µ~è¿™æ˜¯å¤æ‚æ¶ˆæ¯æµ‹è¯•ï¼",
                    {
                        "type": "Text",
                        "data": {"text": "æ–‡æœ¬æ¶ˆæ¯æ®µæµ‹è¯•"}
                    },
                    {
                        "type": "Face",
                        "data": 14
                    }
                ]
            }

        # å‘é€æ¶ˆæ¯
        logger.info(f"\n2. ğŸ“¨ å‘é€æµ‹è¯•æ¶ˆæ¯åˆ°: {args.target_type} {args.target_id}")

        send_result = await client.send_message(
            target_id=args.target_id,
            target_type=args.target_type,
            content_data=content_data,
            sender_id="test_client"
        )

        if send_result and send_result.get('success'):
            logger.info("ğŸ‰ æµ‹è¯•æ¶ˆæ¯å‘é€æˆåŠŸ!")
        else:
            logger.error("âŒ æµ‹è¯•æ¶ˆæ¯å‘é€å¤±è´¥")

        print("\n" + "="*60)
        print("ğŸ¯ æµ‹è¯•å®Œæˆ!")
        print("="*60)

        # æ‰“å°æ€»ç»“
        if send_result:
            print(f"\nğŸ“Š æµ‹è¯•æ€»ç»“:")
            print(f"  ç›®æ ‡: {args.target_type} {args.target_id}")
            print(f"  æ¶ˆæ¯ç±»å‹: {args.content_type}")
            print(f"  å‘é€ç»“æœ: {'âœ… æˆåŠŸ' if send_result.get('success') else 'âŒ å¤±è´¥'}")
            print(f"  å“åº”æ¶ˆæ¯: {send_result.get('message', 'N/A')}")
            print(f"  è¿æ¥æ–¹å¼: Consul å‘ç° ({args.consul} -> {args.service})")

    except KeyboardInterrupt:
        logger.info("\nâŒ¨ï¸ ç”¨æˆ·ä¸­æ–­æµ‹è¯•")
    except Exception as e:
        logger.error(f"ğŸ’¥ æµ‹è¯•è¿‡ç¨‹ä¸­å‘ç”Ÿé”™è¯¯: {e}")
        import traceback
        traceback.print_exc()
    finally:
        await client.close()


if __name__ == "__main__":
    asyncio.run(main())
